File: ncr/normalize.py
"""Normalize model output into the canonical reading-plan schema.

Models don't reliably emit our exact JSON shape — they nest change units inside
chapters, rename `symbol`->`label`, use `overview`->`summary`, omit `file`, etc.
This coerces those common variations into the canonical form (flat `units[]` +
chapters with `nodes[{unit, depth}]`) so the reconciler and renderer work, and
fills each unit's `file`/`symbol` from the block index when the model left them
out. Deterministic; runs before reconcile.
"""

from __future__ import annotations


def _first(d: dict, *keys, default=None):
    for k in keys:
        v = d.get(k)
        if v not in (None, "", []):
            return v
    return default


def normalize_plan(plan: dict, index: dict) -> dict:
    blocks_by_id = {b["blockId"]: b for b in index.get("blocks", [])}

    if not plan.get("overview"):
        plan["overview"] = _first(plan, "overview", "summary", default="")

    units: list[dict] = list(plan.get("units") or [])
    by_id = {u.get("id"): u for u in units if u.get("id")}

    def ingest(raw: dict, fallback_id: str) -> str:
        uid = raw.get("id") or fallback_id
        if uid in by_id:  # already a flat unit; just fill gaps
            _fill(by_id[uid], raw, blocks_by_id)
            return uid
        u = _to_unit(raw, uid, blocks_by_id)
        units.append(u)
        by_id[uid] = u
        return uid

    for ci, ch in enumerate(plan.get("chapters") or []):
        nodes = ch.get("nodes")
        # canonical nodes already reference unit ids -> leave alone
        if nodes and all(isinstance(n, dict) and "unit" in n for n in nodes):
            for n in nodes:
                if n["unit"] in by_id:
                    _fill(by_id[n["unit"]], by_id[n["unit"]], blocks_by_id)
            continue
        inline = _first(ch, "changeUnits", "units", "nodes", default=[])
        ch["nodes"] = [
            {"unit": ingest(cu, f"u-c{ci}-{j}"), "depth": cu.get("depth", 0)}
            for j, cu in enumerate(inline)
        ]
        ch.pop("changeUnits", None)

    # orphans may also carry inline units instead of id references
    norm_orphans = []
    for grp in plan.get("orphans") or []:
        ids = []
        for item in grp.get("units", []):
            if isinstance(item, dict):
                ids.append(ingest(item, f"u-orphan-{len(units)}"))
            else:
                ids.append(item)
        norm_orphans.append({"layer": grp.get("layer"), "units": ids})
    plan["orphans"] = norm_orphans

    # ensure every already-flat unit has file/symbol filled from its blocks
    for u in units:
        _fill(u, u, blocks_by_id)

    plan["units"] = units
    plan.setdefault("edges", [])
    return plan


def _to_unit(raw: dict, uid: str, blocks_by_id: dict) -> dict:
    u = {
        "id": uid,
        "blocks": raw.get("blocks", []),
        "symbol": _first(raw, "symbol", "label", "name", default=""),
        "summary": raw.get("summary", ""),
        "layer": raw.get("layer"),
        "layerReason": _first(raw, "layerReason", "layer_reason", default=""),
        "references": raw.get("references", []),
    }
    if raw.get("detail"):
        u["detail"] = raw["detail"]
    _fill(u, raw, blocks_by_id)
    return u


def _fill(u: dict, raw: dict, blocks_by_id: dict) -> None:
    """Fill file / language / kind from the model or, failing that, the blocks."""
    blk = next((blocks_by_id[b] for b in u.get("blocks", []) if b in blocks_by_id), None)
    if not u.get("file"):
        u["file"] = _first(raw, "file", "path", default=blk["path"] if blk else "")
    if not u.get("language") and raw.get("language"):
        u["language"] = raw["language"]
    if not u.get("symbol"):
        u["symbol"] = _first(raw, "symbol", "label", "name", default="")

File: normalize.go
package main

import (
	"encoding/json"
	"fmt"
)

// Coerce flexible model output into the canonical ReadingPlan — port of
// ncr/normalize.py. Models nest change units in chapters, use label/name for
// symbol, summary for overview, and omit file; this flattens all that and fills
// file from the block index. Runs before reconcile.

func firstNonEmpty(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}

func fillUnit(u *Unit, blocksByID map[string]Block) {
	if u.File == "" {
		for _, b := range u.Blocks {
			if blk, ok := blocksByID[b]; ok {
				u.File = blk.Path
				break
			}
		}
	}
}

func toUnit(ru rawUnit, id string, blocksByID map[string]Block) Unit {
	u := Unit{
		ID:          id,
		Blocks:      ru.Blocks,
		Symbol:      firstNonEmpty(ru.Symbol, ru.Label, ru.Name),
		Summary:     ru.Summary,
		Detail:      ru.Detail,
		Layer:       ru.Layer,
		LayerReason: ru.LayerReason,
		File:        firstNonEmpty(ru.File, ru.Path),
		Language:    ru.Language,
	}
	fillUnit(&u, blocksByID)
	return u
}

func normalizePlan(raw rawPlan, index Index) ReadingPlan {
	blocksByID := map[string]Block{}
	for _, b := range index.Blocks {
		blocksByID[b.BlockID] = b
	}

	var units []Unit
	unitIdx := map[string]int{}
	ensure := func(ru rawUnit, fallbackID string) string {
		id := ru.ID
		if id == "" {
			id = fallbackID
		}
		if idx, ok := unitIdx[id]; ok {
			fillUnit(&units[idx], blocksByID)
			return id
		}
		units = append(units, toUnit(ru, id, blocksByID))
		unitIdx[id] = len(units) - 1
		return id
	}

	for i, ru := range raw.Units { // already-flat units
		ensure(ru, fmt.Sprintf("u-%d", i))
	}

	var chapters []Chapter
	for ci, rc := range raw.Chapters {
		var nodes []Node
		if len(rc.Nodes) > 0 { // canonical node references
			for _, n := range rc.Nodes {
				if idx, ok := unitIdx[n.Unit]; ok {
					fillUnit(&units[idx], blocksByID)
				}
				nodes = append(nodes, n)
			}
		} else {
			inline := rc.ChangeUnits
			if len(inline) == 0 {
				inline = rc.Units
			}
			for j, cu := range inline {
				id := ensure(cu, fmt.Sprintf("u-c%d-%d", ci, j))
				nodes = append(nodes, Node{Unit: id, Depth: cu.Depth})
			}
		}
		chapters = append(chapters, Chapter{ID: rc.ID, Title: rc.Title, Summary: rc.Summary, Nodes: nodes})
	}

	var orphans []Orphan
	for _, ro := range raw.Orphans {
		var ids []string
		for _, item := range ro.Units {
			var s string
			if json.Unmarshal(item, &s) == nil {
				ids = append(ids, s)
				continue
			}
			var ru rawUnit
			if json.Unmarshal(item, &ru) == nil {
				ids = append(ids, ensure(ru, fmt.Sprintf("u-orphan-%d", len(units))))
			}
		}
		orphans = append(orphans, Orphan{Layer: ro.Layer, Units: ids})
	}

	for i := range units {
		fillUnit(&units[i], blocksByID)
	}

	edges := raw.Edges
	if edges == nil {
		edges = []Edge{}
	}
	return ReadingPlan{
		Title:    raw.Title,
		PRNumber: raw.PRNumber,
		Overview: firstNonEmpty(raw.Overview, raw.Summary),
		Chapters: chapters,
		Orphans:  orphans,
		Units:    units,
		Edges:    edges,
	}
}
